### ISSUE TITLE
encoding/json: clarify Decoder.InputOffset semantics

### ISSUE THREAD

**Issue by dsnet** [golang/go#42571]
The documentation on [`Decoder.InputOffset`](https://golang.org/pkg/encoding/json/#Decoder.InputOffset) currently says:
>The offset gives the location of the end of the most recently returned token and the beginning of the next token.

This explanation is somewhat incorrect since the offset cannot be a single value if there are intervening whitespace in-between two JSON tokens. Furthermore, the `json` package implicitly handles colons and commas, which are technically JSON tokens according to RFC 7159, section 2. May the offset ever be past these implicit colons and commas, or will it always be before them? The "token" in this context is a somewhat ambiguous.

The documentation should be clarified to say one of the following:
* The offset may be between the end of the most recently returned token (exclusive), and before the beginning of the next token that may be returned (inclusive).
* The offset points to the next character immediately after the most recently returned token.
* The offset points to the first character of the next token that may be returned.

I chose the phrase "may be returned" to indicate that colons and commas aren't included since they aren't ever returned by the `json` API. I haven't checked the implementation yet to see what it actually returns.

The first definition gives more flexibility to how the `Decoder` may actually be implemented. The latter two definitions gives the user more assurance about what `InputOffset` actually means.

Updates #29688

**Comment by mvdan** [comment]
I agree that the current definition is a bit ambiguous, happy to review a CL. I assume we should just document the current behavior.

**Comment by Windsooon** [comment]
I love the third idea, how about:

    "The offset points to the first character of the next token that may be returned. (except colons or commas)"

The current comments and implement at [src/encoding/json/stream.go](https://cs.opensource.google/go/go/+/refs/tags/go1.17:src/encoding/json/stream.go;l=282;drc=refs%2Ftags%2Fgo1.17) didn't treat colons or commas as Token:

> // A Token holds a value of one of these types:
> //
> //	Delim, for the four JSON delimiters [ ] { }
> //	bool, for JSON booleans
> //	float64, for JSON numbers
> //	Number, for JSON numbers
> //	string, for JSON string literals
> //	nil, for JSON null

It looks like we have to change a lot of comments if we want to fix it.

**Comment by dsnet** [comment]
> "The offset points to the first character of the next token that may be returned. (except colons or commas)"

This is the most user friendly semantic, but breaks in other ways since it requires reading from the input stream until it locates the next token. This would cause the `Decoder` to block on a read (e.g., from a network socket) in cases when it otherwise would not have.

> It looks like we have to change a lot of comments if we want to fix it.

A JSON token is a well defined in [RFC section 2](https://datatracker.ietf.org/doc/html/rfc8259#section-2). I don't think we need to enumerate every possible case.

**Comment by soreavis** [comment]
Reviving this — the doc is still bare at tip, and I reproduced the current behavior to pin down what "document the current behavior" should say, since the thread floated three phrasings and never settled on one, and none of them quite matches what the decoder does:

```
input: [ "a" , "b" ]      // "a" = bytes 2-4, "," = 6, "b" = 8-10
Token() -> "a"   InputOffset() = 5   // end of "a"
More()  -> true  InputOffset() = 6   // the separating comma, NOT "b" (which begins at byte 8)

input: [1,
Token() -> 1     InputOffset() = 2   // end of "1"
Token() -> EOF   InputOffset() = 3   // advanced past the trailing comma, with no More() call
```

So neither half of the present doc ("the end of the most recently returned token *and* the beginning of the next token") holds in general. Without `More` and before `io.EOF`, the offset is the end of the most recently returned token; a `More` call between elements advances only to the separating comma or colon (it lands on a *token* only for the first element of each array or object); and after `Token` returns `io.EOF` the offset can advance past the last token across a consumed separator. That matches exactly why @dsnet rejected the "points to the next token" reading. v1 and the `encoding/json/v2` decoder behave identically here (verified at go1.27-devel).

Since the exact position varies, @dsnet's first suggestion — a range rather than a single point — looks like the accurate fit, with one correction: the lower bound should be inclusive, since `InputOffset` equals the end of the most recently returned token whenever no peek has occurred (`Token(1)` above returns offset 2, the end of `1`). So:

> The offset is between the end of the most recently returned token (inclusive) and the position where the next token would begin; its exact value depends on whether the decoder has peeked ahead (see [Decoder.More]) or reached the end of the input.

If that works I'm happy to send the CL for `stream.go` and `v2_stream.go`. (`encoding/xml`'s `Decoder.InputOffset` carries the identical stale sentence, but its decoder has no `More` and counts per byte, so that's a separate fix with its own wording — I can file it separately.) Or if you'd rather the doc spell out the `More`/EOF cases explicitly instead of a range, I can do that.

**Comment by dsnet** [comment]
Turns out that the v1 `Decoder.InputOffset` is hard to describe as it's not even a stateless method. If you perform some other operations on it, the exact value changes. Regardless, with the release of Go 1.27, we now have `jsontext.Decoder`, for which the `InputOffset` method is clearly defined as pointing to "the next byte immediately after the most recently returned token or value".
